lib/block_orthogonalize.cu

```diff
@@ -1,188 +1,211 @@
-       orthogonalization.
-     */
     template <typename Rotator, typename Vector, std::size_t... S>
-    void CPU(const std::vector<ColorSpinorField*> &B, std::index_sequence<S...>) {
-      typedef BlockOrthoArg<Rotator,Vector,nSpin,spinBlockSize,coarseSpin,nVec> Arg;
-      Arg arg(V, fine_to_coarse, coarse_to_fine, QUDA_INVALID_PARITY, geo_bs, n_block_ortho, V, B[S]...);
-      blockOrthoCPU<sumType,RegType,nSpin,spinBlockSize,nColor,coarseSpin,nVec,Arg>(arg);
+    void launch_host_(const TuneParam &tp, const qudaStream_t &stream,
+                     const std::vector<ColorSpinorField*> &B, std::index_sequence<S...>)
+    {
+      Arg<false, Rotator, Vector> arg(V, fine_to_coarse, coarse_to_fine, QUDA_INVALID_PARITY, geo_bs, n_block_ortho, V, B[S]...);
+      launch_host<BlockOrtho_, OrthoAggregates>(tp, stream, arg);
+      if (two_pass && iter == 0 && V.Precision() < QUDA_SINGLE_PRECISION && !activeTuning()) max = Rotator(V).abs_max(V);
     }
 
-    /**
-       @brief Helper function for expanding the std::vector into a
-       parameter pack that we can use to instantiate the const arrays
-       in BlockOrthoArg and then call the GPU variant of the block
-       orthogonalization.
-     */
     template <typename Rotator, typename Vector, std::size_t... S>
-    void GPU(const TuneParam &tp, const qudaStream_t &stream, const std::vector<ColorSpinorField*> &B, std::index_sequence<S...>) {
-      typedef typename mapper<vFloat>::type RegType; // need to redeclare typedef (WAR for CUDA 7 and 8)
-      typedef BlockOrthoArg<Rotator,Vector,nSpin,spinBlockSize,coarseSpin,nVec> Arg;
-      Arg arg(V, fine_to_coarse, coarse_to_fine, QUDA_INVALID_PARITY, geo_bs, n_block_ortho, V, B[S]...);
-      arg.swizzle = tp.aux.x;
-#ifdef JITIFY
-      using namespace jitify::reflection;
-      auto instance = program->kernel("quda::blockOrthoGPU")
-        .instantiate((int)tp.block.x,Type<sumType>(),Type<RegType>(),nSpin,spinBlockSize,nColor,coarseSpin,nVec,Type<Arg>());
-      cuMemcpyHtoDAsync(instance.get_constant_ptr("quda::B_array_d"), B_array_h, MAX_MATRIX_SIZE, stream);
-      jitify_error = instance.configure(tp.grid,tp.block,tp.shared_bytes,stream).launch(arg);
-#else
-      cudaMemcpyToSymbolAsync(B_array_d, B_array_h, MAX_MATRIX_SIZE, 0, cudaMemcpyHostToDevice, stream);
-      LAUNCH_KERNEL_MG_BLOCK_SIZE(blockOrthoGPU,tp,stream,arg,sumType,RegType,nSpin,spinBlockSize,nColor,coarseSpin,nVec,Arg);
-#endif
+    void launch_device_(const TuneParam &tp, const qudaStream_t &stream,
+                        const std::vector<ColorSpinorField*> &B, std::index_sequence<S...>)
+    {
+      Arg<true, Rotator, Vector> arg(V, fine_to_coarse, coarse_to_fine, QUDA_INVALID_PARITY, geo_bs, n_block_ortho, V, B[S]...);
+      arg.swizzle_factor = tp.aux.x;
+      launch_device<BlockOrtho_, OrthoAggregates>(tp, stream, arg);
+      if (two_pass && iter == 0 && V.Precision() < QUDA_SINGLE_PRECISION && !activeTuning()) max = Rotator(V).abs_max(V);
```

Comment:
Can this line (and the analogous line in `launch_host_`) get moved into the constructor, within the `if` on line 117? I lost some time figuring out where `max` was updated. If there's a reason why it can't be, can you please add a comment around line 117 noting that `max` is updated in the `launch_*_` functions?

Reply:
We cannot, unfortunately.  Since the `abs_max` max function requires the accessor, which only exists in the launch functions, we have to put the max setting where it is.